Sync/BatchingClient.swift

```diff
@@ -0,0 +1,230 @@
+    private var onCollectionUploaded: (POSTResult -> DeferredTimestamp)
+
+    // TODO: use I-U-S.
+    // Each time we do the storage operation, we might receive a backoff notification.
+    // For a success response, this will be on the subsequent request, which means we don't
+    // have to worry about handling successes and failures mixed with backoffs here.
+
+    init(config: InfoConfiguration, ifUnmodifiedSince: Timestamp? = nil, serializeRecord: (Record<T>) -> String?,
+         uploader: BatchUploadFunction, onCollectionUploaded: (POSTResult -> DeferredTimestamp))
+    {
+        self.config = config
+        self.ifUnmodifiedSince = ifUnmodifiedSince
+        self.uploader = uploader
+        self.serializeRecord = serializeRecord
+
+        self.batchMeta = BatchMeta(config: config)
+        self.currentPayload = Payload(config: config)
+        self.onCollectionUploaded = onCollectionUploaded
+    }
+
+    public func addRecords(records: [Record<T>]) -> Success {
+        return walk(records, f: addRecord)
+    }
+
+    public func addRecord(record: Record<T>) -> Success {
+        guard let line = self.serializeRecord(record) else {
+            return deferMaybe(SerializeRecordFailure(record: record))
+        }
+
+        let lineSize = line.utf8.count
+        guard lineSize < Sync15StorageClient.maxRecordSizeBytes else {
+            return deferMaybe(RecordTooLargeError(size: lineSize, guid: record.id))
+        }
+
+        let uploadRecord: UploadRecord = (record.id, line, lineSize)
+
+        // If we can add a record to the payload, all is good - we don't need to push to the server yet.
+        // Otherwise, try to push the payload (since it's full) to the server in a batch
+        if self.currentPayload.add(uploadRecord) {
+            return succeed()
+        }
+
+        // If we have a batch token, go ahead and push up the payload using the token.
+        // Otherwise, attempt to start a new batch. If we fail starting - no problem! The records would
+        // have already been sent in a single POST call.
+        guard let token = batchToken else {
+            return self.start(self.currentPayload)
+                >>== effect({ self.batchToken = $0 })
+
+                // Based on the precondition, we must be able to add at least one payload into the batch 
+                // without it failing
+                >>> effect({
+                    self.batchMeta.add(self.currentPayload)
+                    self.currentPayload = self.addRecordToNewPayload(uploadRecord)
+                })
+                >>> succeed
+        }
+
+        // If we can fit in another payload, push it up using our batch token
+        if batchMeta.add(self.currentPayload) {
+            return self.uploadPayload(self.currentPayload, queryParams: [NSURLQueryItem(name: "batch", value: token)])
+                >>> effect({
+                    self.currentPayload = self.addRecordToNewPayload(uploadRecord)
```

Comment:
This really confirms my feeling that one spot should be responsible for atomically swapping out a new payload. There should be only two ways to get a new payload:

- When an upload begins.
- When a batch boundary is hit.